cmd/datasource-token-refresh/README.md

```diff
@@ -0,0 +1,58 @@
+# Datasource Token Refresh
+
+This binary refreshes a Grafana Prometheus datasource HTTP headers with new oAuth2 access token. This allows you to directly query Google Cloud Monitoring with Grafana.
+
+(Google access tokens have a lifetime of 1 hour.)[https://cloud.google.com/docs/authentication/token-types#at-lifetime] This script should be run hourly to ensure you have an uninterrupted connection between Grafana and Google Cloud Monitoring.
+
+### Run
+
+1. [Make sure you have setup your Grafana datasource setup.](https://cloud.google.com/stackdriver/docs/managed-prometheus/query#grafana-datasource). Extract datasource IDs you wish to use from URL.
+
+2. [Setup a Grafana service account](https://grafana.com/docs/grafana/latest/administration/service-accounts/#create-a-service-account-in-grafana) and generate a token.
+
+3. Figure out the URL of your Grafana instance. e.g. "https://your.grafana.net"
+
+Running the following command should update your datasources with a new token:
+
+```bash
+GRAFANA_TOKEN=YOUR_GRAFANA_SERVICE_ACCOUNT_TOKEN
+GRAFANA_ENDPOINT=YOUR_GRAFANA_INSTANCE_URL
+DATASOURCE_UIDS=YOUR_DATASOURCE_UIDs
+```
+```bash
+go run main.go \
+  --grafana-token=$GRAFANA_TOKEN \
+  --grafana-endpoint=$GRAFANA_ENDPOINT \
+  # Can be a comma separated list of datasource IDs
+  --datasource-uids=$DATASOURCE_UIDS
+```
+### Kubernetes Cron Job
+
+You can run a cron job that runs every 30 minutes to refresh your datasources with a new token using datasource-token-refresh.yaml:
+
+```bash
+cat datasource-token-refresh.yaml \
+| sed 's|$GRAFANA_TOKEN|'"$GRAFANA_TOKEN"'|; s|$GRAFANA_ENDPOINT|'"$GRAFANA_ENDPOINT"'|; s|$DATASOURCE_UIDS|'"$DATASOURCE_UIDS"'|' \
+| kubectl apply -f -
+```
+
+### Custom Google Cloud Monitoring Service Account and Monioring Endpoint
+
+To use with a custom Google Cloud Monitoring service account and ensure your datasources GCM target is setup correctly, you can use the following command:
+
+```bash
+PROJECT_ID=$(gcloud config get-value core/project)
```

Comment:
We should ask the user to manually put in the project ID of the metrics scope they want to use, instead of inferring it from gcloud.

Adding a little friction here is good; we WANT the user to set up a multi-project metrics scope and then query that, if we make it too easy they will accidentally use their local project which is likely not a scope.

Reply:
Thinking more about this - we will want to have a way for users to override project_id so that they can run this in any random project, configure the container for the scoping project, give their random project default svc account access to the scoping project, and have it work. We don't want to require making infrastructure in a scoping project.